File: python/solutions/solution_base.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Callable, TypeVar

from models import Grid, RunType, SolutionOptions, SolutionType
from services import IFileLoader

from .isolution import ISolution

T = TypeVar("T")


class SolutionBase(ISolution, ABC):
    def __init__(self, file_loader: IFileLoader, options: SolutionOptions) -> None:
        self._file_loader = file_loader
        self._options = options

    @property
    def current_options(self) -> SolutionOptions:
        return self._options

    def load_raw(self) -> str:
        return self._file_loader.load_raw(
            self._options.day, self._options.solution_type, self._options.run_type
        )

    def load_lines(self, converter: Callable[[str], T] | None = None) -> list[T]:
        return self._file_loader.load_lines(
            self._options.day,
            self._options.solution_type,
            self._options.run_type,
            converter,
        )

    def load_grid(self, converter: Callable[[str], T] | None = None) -> Grid[T]:
        return self._file_loader.load_grid(
            self._options.day,
            self._options.solution_type,
            self._options.run_type,
            converter,
        )

    def load_items(self, converter: Callable[[str], T] | None = None) -> list[list[T]]:
        return self._file_loader.load_items(
            self._options.day,
            self._options.solution_type,
            self._options.run_type,
            converter,
        )

    @property
    @abstractmethod
    def day(self) -> int:
        ...

    @abstractmethod
    def solve(self) -> str:
        ...

File: dotnet/Solutions/SolutionBase.cs
using AdventOfCode.Models;
using AdventOfCode.Services;
using Microsoft.Extensions.Options;

namespace AdventOfCode.Solutions;

public abstract class SolutionBase(IFileLoader fileLoader, IOptions<SolutionOptions> options) : ISolution
{
    protected SolutionOptions CurrentOptions => options.Value;

    protected Task<T> LoadRaw<T>()
    {
        var (day, solutionType, runType) = ResolveContext();
        return fileLoader.LoadRaw<T>(day, solutionType, runType);
    }

    protected Task<List<T>> LoadLines<T>()
    {
        var (day, solutionType, runType) = ResolveContext();
        return fileLoader.LoadLines<T>(day, solutionType, runType);
    }

    protected Task<Grid<T>> LoadGrid<T>()
    {
        var (day, solutionType, runType) = ResolveContext();
        return fileLoader.LoadGrid<T>(day, solutionType, runType);
    }

    protected Task<List<List<T>>> LoadItems<T>()
    {
        var (day, solutionType, runType) = ResolveContext();
        return fileLoader.LoadItems<T>(day, solutionType, runType);
    }

    private (int Day, SolutionType SolutionType, RunType RunType) ResolveContext()
    {
        var options = CurrentOptions;
        return (options.Day, options.SolutionType, options.RunType);
    }

    public abstract int Day { get; }
    public abstract Task<string> Solve();
}